processor/src/host/handlers.rs

```diff
@@ -0,0 +1,166 @@
+use alloc::{
+    boxed::Box,
+    collections::{BTreeMap, btree_map::Entry},
+    vec::Vec,
+};
+use core::{error::Error, fmt, fmt::Debug};
+
+use miden_core::DebugOptions;
+
+use crate::{ExecutionError, ProcessState};
+
+// EVENT HANDLER TRAIT
+// ================================================================================================
+
+/// An [`EventHandler`] defines a function that that can be called from the processor, which can
+/// read the current execution state and modify the advice.
+///
+/// A struct implementing this trait can access its own state, but any output it produces must
+/// be stored in the process's advice provider.
+pub trait EventHandler: Send + Sync + 'static {
+    /// Handles the event when triggered.
+    fn on_event(&self, process: &mut ProcessState) -> Result<(), HandlerError>;
+}
```

Comment:
Just for my understanding: Each host implementation would basically use the `EventHandlerRegistry` as a convenient way to build up event handlers and then delegate to them when `Host::on_event` is called.

If handling of an event would require asynchronicity, then that would have to be handled separately from the registry, since `EventHandler::on_event` does not return a future, right?

Reply:
Indeed, now that we have `AsyncHost` and `SyncHost`, I feel like this `EventHandler` only makes sense to be used with `SyncHost`, and we'd need an analogous `AsyncEventHandler`?

Reply:
Just to echo my comment from https://github.com/0xMiden/miden-vm/issues/1994: my original intent for these handlers was to be simple functions which don't require any `async` functionality. Basically, these functions would be able to examine the process state and update the advice provider, but would not be able to get "external" data. For this, we'd use the host's `on_event()` handler, which (hopefully), we can figure out how to make `async`.